stock_custom_exchange_rate_date/tests/test_stock.py

```diff
@@ -0,0 +1,154 @@
+        )
+        stock_journal = cls.env["account.journal"].create({"name": "Stock journal", "type": "sale", "code": "STK00"})
+
+        # Creating a new product to ensure valuation is clean
+        cls.stock_account_product_categ = cls.env["product.category"].create(
+            {
+                "name": "Test category",
+                "property_valuation": "real_time",
+                "property_cost_method": "fifo",
+                "property_account_income_categ_id": account_income.id,
+                "property_account_expense_categ_id": account_expense.id,
+                "property_stock_account_input_categ_id": account.id,
+                "property_stock_account_output_categ_id": account_output.id,
+                "property_stock_valuation_account_id": account_valuation.id,
+                "property_stock_journal": stock_journal.id,
+            }
+        )
+        cls.product = cls.env["product.product"].create(
+            {
+                "name": "Product fifo",
+                "type": "product",
+                "default_code": "PR-FIFO",
+                "categ_id": cls.stock_account_product_categ.id,
+            }
+        )
+
+    def create_purchase_order(self, partner=None, **line_kwargs):
+        if partner is None:
+            partner = self.vendor
+        purchase_order = Form(self.env["purchase.order"])
+        purchase_order.partner_id = partner
+        purchase_order.currency_id = self.eur
+        purchase_order = purchase_order.save()
+        self.create_po_line(purchase_order, **line_kwargs)
+        return purchase_order
+
+    def create_po_line(self, purchase_order, product=None, quantity=1, price=100):
+        if product is None:
+            product = self.product
+        with Form(purchase_order) as po:
+            with po.order_line.new() as line:
+                line.product_id = product
+                line.product_qty = quantity
+                line.price_unit = price
+
+    def set_currency_rates(self, rate_date, usd_rate, eur_rate):
+        # Remove existing rates, if any
+        rate_model = self.env["res.currency.rate"]
+        current_rates = rate_model.search(
+            [
+                ("name", "=", rate_date),
+                ("currency_id", "in", [self.usd.id, self.eur.id]),
+            ]
+        )
+        current_rates.unlink()
+
+        # Create new rates
+        rate_model.create(
+            {
+                "currency_id": self.usd.id,
+                "rate": usd_rate,
+                "name": rate_date,
+            }
+        )
+        rate_model.create({"currency_id": self.eur.id, "rate": eur_rate, "name": rate_date})
+
+    def test_01_date_rate_set(self):
+        """Set a custom rate date on a transfer, valuation should be computed using that date"""
+        # Purchase product
+        self.set_currency_rates(rate_date=self.today, usd_rate=1, eur_rate=1.25)
```

Comment:
Why don't you set them in the `setUpClass` method?

Reply:
Because I had errors calling it from the `setUpClass` method.

Reply:
@keylor2906 please don't resolve threads yourself, because I'm not able which of them were actually addressed or what ones would need further attention.